zkevm-circuits/src/pi_circuit.rs

```diff
@@ -658,54 +687,116 @@ impl<F: Field, const MAX_TXS: usize, const MAX_CALLDATA: usize>
-    pub fn new(randomness: impl Into<F>, rand_rpi: impl Into<F>, public_data: PublicData) -> Self {
+    pub fn new(
+        max_txs: usize,
+        max_calldata: usize,
+        randomness: impl Into<F>,
+        rand_rpi: impl Into<F>,
+        public_data: PublicData,
+    ) -> Self {
         Self {
+            max_txs,
+            max_calldata,
             randomness: randomness.into(),
             rand_rpi: rand_rpi.into(),
             public_data,
         }
     }
 }
-impl<F: Field, const MAX_TXS: usize, const MAX_CALLDATA: usize> Circuit<F>
-    for PiCircuit<F, MAX_TXS, MAX_CALLDATA>
-{
-    type Config = PiCircuitConfig<F, MAX_TXS, MAX_CALLDATA>;
-    type FloorPlanner = SimpleFloorPlanner;
 
-    fn without_witnesses(&self) -> Self {
-        Self::default()
+impl<F: Field> SubCircuit<F> for PiCircuit<F> {
+    type Config = PiCircuitConfig<F>;
+
+    fn new_from_block(block: &witness::Block<F>) -> Self {
+        let public_data = PublicData {
+            chain_id: block.context.chain_id,
+            history_hashes: block.context.history_hashes.clone(),
+            transactions: block.eth_block.transactions.clone(),
+            state_root: block.eth_block.state_root,
+            prev_state_root: H256::from_uint(&block.prev_state_root),
+            block_constants: BlockConstants {
+                coinbase: block.context.coinbase,
+                timestamp: block.context.timestamp,
+                number: block.context.number.as_u64().into(),
+                difficulty: block.context.difficulty,
+                gas_limit: block.context.gas_limit.into(),
+                base_fee: block.context.base_fee,
+            },
+        };
+        PiCircuit::new(
+            block.circuits_params.max_txs,
+            block.circuits_params.max_calldata,
+            block.randomness,
+            block.randomness + F::from_u128(1),
+            public_data,
+        )
     }
 
-    fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
-        let block_table = BlockTable::construct(meta);
-        let tx_table = TxTable::construct(meta);
-        PiCircuitConfig::new(meta, block_table, tx_table)
+    /// Compute the public inputs for this circuit.
+    fn instance(&self) -> Vec<Vec<F>> {
+        let rlc_rpi_col = raw_public_inputs_col::<F>(
+            self.max_txs,
+            self.max_calldata,
+            &self.public_data,
+            self.randomness,
+        );
+        assert_eq!(
+            rlc_rpi_col.len(),
+            BLOCK_LEN + 1 + EXTRA_LEN + 3 * (TX_LEN * self.max_txs + 1 + self.max_calldata)
+        );
+
+        // Computation of raw_pulic_inputs
+        let rlc_rpi = rlc_rpi_col
+            .iter()
+            .rev()
+            .fold(F::zero(), |acc, val| acc * self.rand_rpi + val);
+
+        // let block_hash = public_data
+        //     .eth_block
+        //     .hash
+        //     .unwrap_or_else(H256::zero)
+        //     .to_fixed_bytes();
```

Comment:
Ah, I just moved this code around, this was already there: https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/43da558aac7178f1895925782874662d26a72642/zkevm-circuits/src/pi_circuit.rs#L863-L867

I guess this will be uncommented once we validate the block hash.

Reply:
I see, make sense to keep it.